StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift

```diff
@@ -1475,13 +1475,25 @@ public class STPPaymentHandler: NSObject {
         return resultingUrl
     }
 
-    func _retryAfterDelay(retryCount: Int, delayTime: TimeInterval = 3, block: @escaping STPVoidBlock) {
-        DispatchQueue.main.asyncAfter(deadline: .now() + delayTime) {
+    /// Schedules delayed execution of a retry block and records a poll attempt.
+    /// - Parameters:
+    ///   - delay: Time interval to wait before execution (used as target interval for optimization)
+    ///   - pollingBudget: Budget tracker for polling attempts
+    ///   - block: Block to execute if budget allows
+    func pollAfterDelay(delay: TimeInterval = 1, pollingBudget: PollingBudget, block: @escaping () -> Void) {
+        let optimizedDelay = pollingBudget.recommendedDelay(targetInterval: delay)
+        DispatchQueue.main.asyncAfter(deadline: .now() + optimizedDelay) {
+            pollingBudget.recordPollAttempt()
             block()
         }
     }
 
-    func _retrieveAndCheckIntentForCurrentAction(currentAction: STPPaymentHandlerActionParams? = nil, retryCount: Int = maxChallengeRetries) {
+    /// Retrieves and checks the payment intent status for the current action.
+    /// If pollingBudget is nil, this is the first attempt and a new budget is created.
+    /// - Parameters:
+    ///   - currentAction: Action parameters to process, defaults to self.currentAction
+    ///   - pollingBudget: Existing polling budget, or nil for first attempt
+    func _retrieveAndCheckIntentForCurrentAction(currentAction: STPPaymentHandlerActionParams? = nil, pollingBudget: PollingBudget? = nil) {
```

Comment:
What happens if the network request times out? Do we enter into the error handling code and stop polling?

Reply:
The behavior when the network timeout is unchanged from master. If the network request times out we retry if the budget allows, if it does not `canPoll` will be false in `_retrieveAndCheckIntentForCurrentAction`, and we will enter into the same error handling as before.